<a href="https://colab.research.google.com/github/Manya-65/assigment1/blob/main/Association_rules.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Analyze the "Online retail.xlsx" dataset using association rule mining to discover relationships between products, interpret the findings, and answer interview questions about association rules.

## Load the dataset

### Subtask:
Load the "Online retail.xlsx" file into a pandas DataFrame.


**Reasoning**:
Import the pandas library, read the Excel file into a DataFrame, and display the first few rows.



In [ ]:
import pandas as pd

df = pd.read_excel("Online retail.xlsx")
display(df.head())

shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
0                             burgers,meatballs,eggs                                                                                                                                                                             
1                                            chutney                                                                                                                                                                             
2                                     turkey,avocado                                                                                                                                                                             
3  mineral water,milk,energy bar,whole wheat rice...                                                                                                                                                                             
4                                     low fat yogurt

## Data preprocessing

### Subtask:
Handle missing values, remove duplicate rows, and convert the data to a suitable format for association rule mining (e.g., one-hot encoded transaction data).


**Reasoning**:
Inspect the DataFrame for missing values and handle them, then identify and remove duplicate rows, and finally convert the data to a suitable format for association rule mining by splitting the comma-separated string and creating a list of lists.



In [ ]:
df.dropna(inplace=True)
df.drop_duplicates(inplace=True)

transactions = df.iloc[:, 0].apply(lambda x: x.split(',')).tolist()

print(f"Number of transactions after cleaning: {len(transactions)}")
print("First 5 transactions after cleaning:")
for i in range(min(5, len(transactions))):
    print(transactions[i])

Number of transactions after cleaning: 5175
First 5 transactions after cleaning:
['burgers', 'meatballs', 'eggs']
['chutney']
['turkey', 'avocado']
['mineral water', 'milk', 'energy bar', 'whole wheat rice', 'green tea']
['low fat yogurt']


## Implement apriori algorithm

### Subtask:
Use the `mlxtend` library to implement the Apriori algorithm.


**Reasoning**:
Implement the Apriori algorithm using mlxtend by one-hot encoding the transactions and converting the result to a DataFrame.



In [ ]:
from mlxtend.preprocessing import TransactionEncoder

te = TransactionEncoder()
te_ary = te.fit(transactions).transform(transactions)
df_encoded = pd.DataFrame(te_ary, columns=te.columns_)
display(df_encoded.head())

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0       False    False              False      False    False        False   
1       False    False              False      False    False        False   
2       False    False              False      False     True        False   
3       False    False              False      False    False        False   
4       False    False              False      False    False        False   

   bacon  barbecue sauce  black tea  blueberries  ...  turkey  vegetables mix  \
0  False           False      False        False  ...   False           False   
1  False           False      False        False  ...   False           False   
2  False           False      False        False  ...    True           False   
3  False           False      False        False  ...   False           False   
4  False           False      False        False  ...   False           False   

   water spray  white wine  whole weat flour  whole wheat pasta  \
0        False       False             False              False   
1        False       False             False              False   
2        False       False             False              False   
3        False       False             False              False   
4        False       False             False              False   

   whole wheat rice   yams  yogurt cake  zucchini  
0             False  False        False     False  
1             False  False        False     False  
2             False  False        False     False  
3              True  False        False     False  
4             False  False        False     False  

[5 rows x 120 columns]

## Apply association rule mining

### Subtask:
Apply the Apriori algorithm to the pre-processed data to find frequent itemsets and generate association rules.


**Reasoning**:
Apply the Apriori algorithm to find frequent itemsets and then generate association rules, storing the rules in a DataFrame.



In [ ]:
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

frequent_itemsets = apriori(df_encoded, min_support=0.01, use_colnames=True)
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.5)
display(rules.head())

antecedents      consequents  antecedent support  \
0              (chocolate, chicken)  (mineral water)            0.021256   
1            (chocolate, olive oil)  (mineral water)            0.023575   
2               (eggs, ground beef)  (mineral water)            0.028792   
3  (frozen vegetables, ground beef)  (mineral water)            0.024541   
4  (frozen vegetables, ground beef)      (spaghetti)            0.024541   

   consequent support   support  confidence      lift  representativity  \
0            0.299710  0.011014    0.518182  1.728943               1.0   
1            0.299710  0.011981    0.508197  1.695627               1.0   
2            0.299710  0.014493    0.503356  1.679475               1.0   
3            0.299710  0.013333    0.543307  1.812775               1.0   
4            0.229565  0.012560    0.511811  2.229480               1.0   

   leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
0  0.004644    1.453432       0.430768  0.035536   0.311973    0.277466  
1  0.004915    1.423923       0.420153  0.038485   0.297715    0.274085  
2  0.005863    1.410043       0.416570  0.046154   0.290802    0.275856  
3  0.005978    1.533393       0.459640  0.042884   0.347852    0.293897  
4  0.006927    1.578149       0.565339  0.052000   0.366346    0.283262

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)


## Analyze and interpret rules

### Subtask:
Analyze the generated rules based on support, confidence, and lift, and interpret the findings to understand customer purchasing behavior.


**Reasoning**:
Sort the rules by lift and display the top 10 to identify the strongest associations and prepare for interpretation.



In [ ]:
rules_sorted = rules.sort_values(by='lift', ascending=False)
display(rules_sorted.head(10))

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

antecedents      consequents  antecedent support  \
4  (frozen vegetables, ground beef)      (spaghetti)            0.024541   
8                      (milk, soup)  (mineral water)            0.021449   
3  (frozen vegetables, ground beef)  (mineral water)            0.024541   
9                 (soup, spaghetti)  (mineral water)            0.020676   
6           (ground beef, pancakes)  (mineral water)            0.020870   
0              (chocolate, chicken)  (mineral water)            0.021256   
7                 (milk, olive oil)  (mineral water)            0.024155   
1            (chocolate, olive oil)  (mineral water)            0.023575   
5               (ground beef, milk)  (mineral water)            0.031691   
2               (eggs, ground beef)  (mineral water)            0.028792   

   consequent support   support  confidence      lift  representativity  \
4            0.229565  0.012560    0.511811  2.229480               1.0   
8            0.299710  0.012367    0.576577  1.923781               1.0   
3            0.299710  0.013333    0.543307  1.812775               1.0   
9            0.299710  0.010821    0.523364  1.746235               1.0   
6            0.299710  0.010821    0.518519  1.730067               1.0   
0            0.299710  0.011014    0.518182  1.728943               1.0   
7            0.299710  0.012367    0.512000  1.708317               1.0   
1            0.299710  0.011981    0.508197  1.695627               1.0   
5            0.299710  0.016039    0.506098  1.688623               1.0   
2            0.299710  0.014493    0.503356  1.679475               1.0   

   leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
4  0.006927    1.578149       0.565339  0.052000   0.366346    0.283262  
8  0.005939    1.653876       0.490716  0.040050   0.395360    0.308920  
3  0.005978    1.533393       0.459640  0.042884   0.347852    0.293897  
9  0.004624    1.469236       0.436362  0.034956   0.319374    0.279735  
6  0.004566    1.454448       0.430982  0.034934   0.312454    0.277312  
0  0.004644    1.453432       0.430768  0.035536   0.311973    0.277466  
7  0.005128    1.435020       0.424892  0.039702   0.303146    0.276632  
1  0.004915    1.423923       0.420153  0.038485   0.297715    0.274085  
5  0.006541    1.417871       0.421148  0.050858   0.294717    0.279806  
2  0.005863    1.410043       0.416570  0.046154   0.290802    0.275856

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

## Answer interview questions

### Subtask:
Provide answers to the interview questions about lift, support, confidence, and limitations of association rule mining.


**Reasoning**:
Provide clear definitions and explanations for lift, support, confidence, and discuss the limitations of association rule mining based on the subtask instructions and the context of the analysis performed so far.



In [ ]:
print("Interview Questions about Association Rule Mining")
print("-" * 50)

# 1. Define lift and explain its values
print("1. Definition and Interpretation of Lift:")
print("   Lift is a metric used to evaluate the strength of an association rule. It measures how much more likely the consequent is to be purchased given that the antecedent has been purchased, compared to the baseline likelihood of purchasing the consequent alone.")
print("\n   - Lift > 1: Indicates a positive correlation between the antecedent and the consequent. The items in the antecedent and consequent are more likely to be purchased together than individually.")
print("   - Lift = 1: Indicates no correlation between the antecedent and the consequent. The likelihood of purchasing the consequent is independent of purchasing the antecedent.")
print("   - Lift < 1: Indicates a negative correlation between the antecedent and the consequent. The items in the antecedent and consequent are less likely to be purchased together than individually.")

print("\n" + "-" * 50)

# 2. Define support and confidence
print("2. Definition of Support and Confidence:")
print("   - Support: Measures the overall popularity of an itemset (a combination of items). For an association rule A -> B, the support is the proportion of transactions in the dataset that contain both A and B.")
print("     Support(A -> B) = (Number of transactions containing A and B) / (Total number of transactions)")
print("\n   - Confidence: Measures the likelihood that the consequent will be purchased when the antecedent has been purchased. For an association rule A -> B, the confidence is the proportion of transactions containing A that also contain B.")
print("     Confidence(A -> B) = (Number of transactions containing A and B) / (Number of transactions containing A)")
print("     Confidence(A -> B) = Support(A -> B) / Support(A)")

print("\n" + "-" * 50)

# 3. Discuss limitations of association rule mining
print("3. Limitations of Association Rule Mining:")
print("   - Scalability: As the number of items and transactions increases, the number of possible itemsets and association rules grows exponentially. This can lead to computational challenges in terms of processing time and memory requirements.")
print("   - Finding Trivial Rules: The algorithm can generate a large number of rules, many of which may be trivial or obvious (e.g., 'bread -> milk' in a grocery store). Filtering and interpreting these rules to find actionable insights can be time-consuming.")
print("   - Handling Continuous Data: Association rule mining algorithms are primarily designed for categorical data. Applying them to continuous data requires discretization, which can lead to loss of information and potentially impact the quality of the generated rules.")
print("   - Interpreting Complex Rules: Rules with multiple items in the antecedent or consequent can be challenging to interpret and understand in a practical business context.")

print("\n" + "-" * 50)

Interview Questions about Association Rule Mining
--------------------------------------------------
1. Definition and Interpretation of Lift:
   Lift is a metric used to evaluate the strength of an association rule. It measures how much more likely the consequent is to be purchased given that the antecedent has been purchased, compared to the baseline likelihood of purchasing the consequent alone.

   - Lift > 1: Indicates a positive correlation between the antecedent and the consequent. The items in the antecedent and consequent are more likely to be purchased together than individually.
   - Lift = 1: Indicates no correlation between the antecedent and the consequent. The likelihood of purchasing the consequent is independent of purchasing the antecedent.
   - Lift < 1: Indicates a negative correlation between the antecedent and the consequent. The items in the antecedent and consequent are less likely to be purchased together than individually.

------------------------------------

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

## Summary:

### Data Analysis Key Findings

*   The dataset was successfully loaded and preprocessed, resulting in 5175 cleaned transactions.
*   The data was transformed into a one-hot encoded format suitable for the Apriori algorithm.
*   The Apriori algorithm identified frequent itemsets with a minimum support of 0.01.
*   Association rules were generated with a minimum confidence of 0.5.
*   Analysis of the top rules sorted by lift revealed strong positive associations between certain product combinations, such as "frozen vegetables" and "ground beef" with "spaghetti", and various items with "mineral water".
*   The definitions and interpretations of lift, support, and confidence were provided.
*   Key limitations of association rule mining, including scalability, handling trivial rules, continuous data, and rule complexity, were discussed.

### Insights or Next Steps

*   Leverage the discovered high-lift association rules for targeted marketing campaigns, product bundling, and optimized store layouts or website recommendations.
*   Explore different support and confidence thresholds to identify a wider range of potentially valuable association rules, and consider implementing metrics beyond lift, support, and confidence for a more comprehensive evaluation of rule significance.
